File: csharp/Matmul.cs
using System.Threading.Tasks;

public abstract class MatmulBase : Benchmark
{
    protected int _n;
    protected uint _result;
    protected double[][] _a;
    protected double[][] _b;

    protected MatmulBase(string typeName)
    {
        TypeName = typeName;
        _result = 0;
    }

    public override uint Checksum => _result;
    public override string TypeName { get; }

    public override void Prepare()
    {
        _n = (int)ConfigVal("n");
        _a = MatGen(_n);
        _b = MatGen(_n);
        _result = 0;
    }

    protected double[][] MatGen(int n)
    {
        double tmp = 1.0 / n / n;
        double[][] a = new double[n][];

        for (int i = 0; i < n; i++)
        {
            a[i] = new double[n];
            for (int j = 0; j < n; j++)
                a[i][j] = tmp * (i - j) * (i + j);
        }

        return a;
    }
}

public class Matmul1T : MatmulBase
{
    public Matmul1T() : base("Matmul::Single") { }

    private double[][] MatMul(double[][] a, double[][] b)
    {
        int n = a.Length;

        double[][] b2 = new double[n][];
        for (int i = 0; i < n; i++)
        {
            b2[i] = new double[n];
            for (int j = 0; j < n; j++)
                b2[i][j] = b[j][i];
        }

        double[][] c = new double[n][];
        for (int i = 0; i < n; i++)
        {
            c[i] = new double[n];
            double[] ai = a[i];
            double[] ci = c[i];

            for (int j = 0; j < n; j++)
            {
                double[] b2j = b2[j];
                double s = 0.0;

                for (int k = 0; k < n; k++)
                    s += ai[k] * b2j[k];

                ci[j] = s;
            }
        }

        return c;
    }

    public override void Run(long iterationId)
    {
        double[][] c = MatMul(_a, _b);
        double value = c[_n >> 1][_n >> 1];
        _result += Helper.Checksum(value);
    }
}

public abstract class MatmulParallel : MatmulBase
{
    protected int _maxDegreeOfParallelism;

    protected MatmulParallel(string typeName, int maxDegreeOfParallelism) : base(typeName)
    {
        _maxDegreeOfParallelism = maxDegreeOfParallelism;
    }

    protected double[][] MatMulParallel(double[][] a, double[][] b)
    {
        int n = a.Length;

        double[][] bT = new double[n][];
        for (int i = 0; i < n; i++)
        {
            bT[i] = new double[n];
            for (int j = 0; j < n; j++)
                bT[i][j] = b[j][i];
        }

        double[][] c = new double[n][];
        for (int i = 0; i < n; i++)
            c[i] = new double[n];

        Parallel.For(0, n, new ParallelOptions { MaxDegreeOfParallelism = _maxDegreeOfParallelism }, i =>
        {
            double[] ai = a[i];
            double[] ci = c[i];

            for (int j = 0; j < n; j++)
            {
                double sum = 0.0;
                double[] bTj = bT[j];

                for (int k = 0; k < n; k++)
                    sum += ai[k] * bTj[k];

                ci[j] = sum;
            }
        });

        return c;
    }

    public override void Run(long iterationId)
    {
        double[][] c = MatMulParallel(_a, _b);
        double value = c[_n >> 1][_n >> 1];
        _result += Helper.Checksum(value);
    }
}

public class Matmul4T : MatmulParallel
{
    public Matmul4T() : base("Matmul::T4", 4) { }
}

public class Matmul8T : MatmulParallel
{
    public Matmul8T() : base("Matmul::T8", 8) { }
}

public class Matmul16T : MatmulParallel
{
    public Matmul16T() : base("Matmul::T16", 16) { }
}

File: fsharp/Matmul.fs
namespace Benchmarks

open System
open System.Threading.Tasks

module MatmulCommon =
    let matGen (n: int) =
        let tmp = 1.0 / float n / float n
        Array.init n (fun i ->
            Array.init n (fun j ->
                tmp * float (i - j) * float (i + j)))

    let transpose (b: double[][]) =
        let n = b.Length
        Array.init n (fun i ->
            Array.init n (fun j -> b.[j].[i]))

    let matMulSingleThread (a: double[][]) (b: double[][]) =
        let n = a.Length
        let bT = transpose b

        Array.init n (fun i ->
            let ai = a.[i]
            Array.init n (fun j ->
                let bTj = bT.[j]
                let mutable sum = 0.0

                for k = 0 to n - 1 do
                    sum <- sum + ai.[k] * bTj.[k]
                sum))

    let matMulMultiThread (threads: int) (a: double[][]) (b: double[][]) =
        let n = a.Length
        let bT = transpose b
        let c = Array.init n (fun _ -> Array.zeroCreate<double> n)

        let options = ParallelOptions(MaxDegreeOfParallelism = threads)

        Parallel.For(0, n, options, fun i ->
            let ai = a.[i]
            let ci = c.[i]

            for j = 0 to n - 1 do
                let bTj = bT.[j]
                let mutable sum = 0.0

                for k = 0 to n - 1 do
                    sum <- sum + ai.[k] * bTj.[k]

                ci.[j] <- sum
            ) |> ignore

        c

[<AbstractClass>]
type MatmulBase(name: string) =
    inherit Benchmark()

    let mutable n = 0
    let mutable result = 0u
    let mutable a : double[][] = null
    let mutable b : double[][] = null

    override this.Checksum = result
    override this.Name = name

    override this.Prepare() =
        n <- int (this.ConfigVal("n"))
        a <- MatmulCommon.matGen n
        b <- MatmulCommon.matGen n
        result <- 0u

    member this.N = n
    member this.A = a
    member this.B = b
    member this.Result with get() = result and set(value) = result <- value

type Matmul1T() =
    inherit MatmulBase("Matmul::Single")

    override this.Run(_: int64) =
        let c = MatmulCommon.matMulSingleThread this.A this.B
        let value = c.[this.N >>> 1].[this.N >>> 1]
        this.Result <- this.Result + Helper.Checksum(value)

type Matmul4T() =
    inherit MatmulBase("Matmul::T4")

    override this.Run(_: int64) =
        let c = MatmulCommon.matMulMultiThread 4 this.A this.B
        let value = c.[this.N >>> 1].[this.N >>> 1]
        this.Result <- this.Result + Helper.Checksum(value)

type Matmul8T() =
    inherit MatmulBase("Matmul::T8")

    override this.Run(_: int64) =
        let c = MatmulCommon.matMulMultiThread 8 this.A this.B
        let value = c.[this.N >>> 1].[this.N >>> 1]
        this.Result <- this.Result + Helper.Checksum(value)

type Matmul16T() =
    inherit MatmulBase("Matmul::T16")

    override this.Run(_: int64) =
        let c = MatmulCommon.matMulMultiThread 16 this.A this.B
        let value = c.[this.N >>> 1].[this.N >>> 1]
        this.Result <- this.Result + Helper.Checksum(value)